From 6cdd552c4ec284b74f99a97d514e80bab6fe227c Mon Sep 17 00:00:00 2001
From: Gustavo Duarte <gduarte@redhat.com>
Date: Tue, 2 Sep 2014 21:09:17 -0400
Subject: [powerpc] implement vmemmap_remove_mapping() for BOOK3S

Message-id: <1409692159-32351-3-git-send-email-gduarte@redhat.com>
Patchwork-id: 90691
O-Subject: [RHEL7.1 PATCH BZ 1090174 2/4] powerpc: implement vmemmap_remove_mapping() for BOOK3S
Bugzilla: 1090174
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1090174
Upstream Status: ed5694a8464a133582c632f8081324408bcc486d

commit ed5694a8464a133582c632f8081324408bcc486d
Author: Li Zhong <zhong@linux.vnet.ibm.com>
Date:   Wed Jun 11 16:23:37 2014 +0800

    powerpc: implement vmemmap_remove_mapping() for BOOK3S

    This is to be called in vmemmap_free(), leave the implementation on BOOK3E
    empty as before.

    Signed-off-by: Li Zhong <zhong@linux.vnet.ibm.com>
    Cc: Nathan Fontenot <nfont@linux.vnet.ibm.com>
    Acked-by: Nathan Fontenot <nfont@linux.vnet.ibm.com>
    Signed-off-by: Benjamin Herrenschmidt <benh@kernel.crashing.org>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/mm/hash_utils_64.c b/arch/powerpc/mm/hash_utils_64.c
index 96df393..f41cc57 100644
--- a/arch/powerpc/mm/hash_utils_64.c
+++ b/arch/powerpc/mm/hash_utils_64.c
@@ -239,7 +239,7 @@ int htab_bolt_mapping(unsigned long vstart, unsigned long vend,
 }
 
 #ifdef CONFIG_MEMORY_HOTPLUG
-static int htab_remove_mapping(unsigned long vstart, unsigned long vend,
+int htab_remove_mapping(unsigned long vstart, unsigned long vend,
         int psize, int ssize)
 {
  unsigned long vaddr;
diff --git a/arch/powerpc/mm/init_64.c b/arch/powerpc/mm/init_64.c
index 35731fe..b52611a 100644
--- a/arch/powerpc/mm/init_64.c
+++ b/arch/powerpc/mm/init_64.c
@@ -212,6 +212,13 @@ static void __meminit vmemmap_create_mapping(unsigned long start,
  for (i = 0; i < page_size; i += PAGE_SIZE)
   BUG_ON(map_kernel_page(start + i, phys, flags));
 }
+
+#ifdef CONFIG_MEMORY_HOTPLUG
+static void vmemmap_remove_mapping(unsigned long start,
+       unsigned long page_size)
+{
+}
+#endif
 #else /* CONFIG_PPC_BOOK3E */
 static void __meminit vmemmap_create_mapping(unsigned long start,
           unsigned long page_size,
@@ -223,6 +230,21 @@ static void __meminit vmemmap_create_mapping(unsigned long start,
      mmu_kernel_ssize);
  BUG_ON(mapped < 0);
 }
+
+#ifdef CONFIG_MEMORY_HOTPLUG
+extern int htab_remove_mapping(unsigned long vstart, unsigned long vend,
+   int psize, int ssize);
+
+static void vmemmap_remove_mapping(unsigned long start,
+       unsigned long page_size)
+{
+ int mapped = htab_remove_mapping(start, start + page_size,
+      mmu_vmemmap_psize,
+      mmu_kernel_ssize);
+ BUG_ON(mapped < 0);
+}
+#endif
+
 #endif /* CONFIG_PPC_BOOK3E */
 
 struct vmemmap_backing *vmemmap_list;
-- 
1.7.1